File: android/Reguerta/app/src/main/java/com/reguerta/user/data/commitments/InMemorySeasonalCommitmentRepository.kt
package com.reguerta.user.data.commitments

import com.reguerta.user.domain.commitments.SeasonalCommitment
import com.reguerta.user.domain.commitments.SeasonalCommitmentRepository
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

class InMemorySeasonalCommitmentRepository(
    items: List<SeasonalCommitment> = emptyList(),
) : SeasonalCommitmentRepository {
    private val mutex = Mutex()
    private val commitments = items.associateBy { it.id }.toMutableMap()

    override suspend fun getActiveCommitmentsForUser(userId: String): List<SeasonalCommitment> = mutex.withLock {
        commitments.values
            .filter { it.userId == userId && it.active }
            .sortedWith(compareBy<SeasonalCommitment> { it.seasonKey }.thenBy { it.productId })
    }
}

File: ios/Reguerta/Reguerta/Data/Commitments/InMemorySeasonalCommitmentRepository.swift
import Foundation

actor InMemorySeasonalCommitmentRepository: SeasonalCommitmentRepository {
    private var commitmentsById: [String: SeasonalCommitment]

    init(items: [SeasonalCommitment] = []) {
        self.commitmentsById = Dictionary(uniqueKeysWithValues: items.map { ($0.id, $0) })
    }

    func activeCommitments(userId: String) async -> [SeasonalCommitment] {
        commitmentsById.values
            .filter { $0.userId == userId && $0.active }
            .sorted {
                if $0.seasonKey.localizedCaseInsensitiveCompare($1.seasonKey) != .orderedSame {
                    return $0.seasonKey.localizedCaseInsensitiveCompare($1.seasonKey) == .orderedAscending
                }
                return $0.productId.localizedCaseInsensitiveCompare($1.productId) == .orderedAscending
            }
    }
}
